openpype/hosts/maya/api/workfile_template_builder.py

```diff
@@ -251,6 +257,8 @@ def cleanup_placeholder(self, placeholder, failed):
         cmds.sets(node, addElement=PLACEHOLDER_SET)
         cmds.hide(node)
         cmds.setAttr(node + ".hiddenInOutliner", True)
+        update_instances_frame_range()
+        update_instances_asset_attribute()
```

Comment:
Are we sure this is in the right place. `cleanup_placeholder` sounds like it runs _per placeholder_ instead of at the very end of workfile build. We only want to run this once, right? Not every single time.

I feel like this is probably in the wrong place.

Reply:
Well seen, it is true that it is executed at each placeholder. I think we should move this to [here](https://github.com/ynput/OpenPype/blob/fe58cb3c7cc4520b61c8b73c166f6694c5d29e91/openpype/pipeline/workfile/workfile_template_builder.py#L712) with a `post_populate_scene_placeholders` method but we don't can't call maya commands here, maybe we should implement instance-related commands in `host` class? maybe it doesn't make sense with the new publisher (which I don't know)?
Tell me what would make more sense.

Reply:
I feel like the abstract workfile builder is missing some methods that trigger before the build and after the build. We could expose a `after_build_process()` and `pre_build_process()` so that each DCC could implement its own callbacks that run before or after ALL placeholders. Not sure if the best approach and whether it's really missing - but couldn't quickly find a feature that does that.

Another way could be for the workfile building to emit events `emit_event("workfile.before.build")` and `emit_event("workfile.after.build")` so that any module could attach callbacks to the events using `register_event_callback` - which the host itself could do e.g. on `install()` etc.

Not sure which would be the best approach. @iLLiCiTiT @tokejepsen @antirotor thoughts?